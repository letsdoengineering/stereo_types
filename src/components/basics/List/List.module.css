.list-unstyled {
    list-style: none;
}

.item-spacing {
    margin: var(--standard-indent)
}
