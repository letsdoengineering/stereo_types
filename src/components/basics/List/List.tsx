import React from 'react'
import getClassNames from '../../../utils/get-class-names'
import * as listStyles from './List.module.css'

export type ListProps = {
  /** optional: default to ul if not passed */
  isOrdered?: boolean
  /** optional: default to unstyled if not passed */
  isUnstyled?: boolean
  /** optional: styles a list to appear as links if set - defaults to false */
  isLinks?: boolean
  /** optional: indents the list if set to true */
  isIndented?: boolean
  /** optional: class-names */
  className?: string
  /** children always of type ReactNode */
  children: React.ReactNode
}

/** Text: renders a Text tag of a type based on the tagName prop */
const List: React.FC<
  ListProps & React.AllHTMLAttributes<HTMLUListElement> & React.AllHTMLAttributes<HTMLOListElement>
> = ({
  isOrdered = false,
  isUnstyled = true,
  isLinks = false,
  isIndented = false,
  className,
  children,
}: ListProps) => {
  const Wrapper = `${isOrdered ? 'ol' : 'ul'}` as keyof JSX.IntrinsicElements
  const classNames = getClassNames({
    className,
    conditionalClasses: {
      [`${listStyles.listUnstyled}`]: isUnstyled,
      [listStyles.itemSpacing]: isIndented,
    },
  })

  const childrenWithProps = React.Children.map(children, (child) => {
    // Checking isValidElement is the safe way and avoids a typescript error too.
    if (React.isValidElement(child)) {
      return React.cloneElement(child, { isLinks })
    }
    return child
  })

  return <Wrapper className={classNames}>{childrenWithProps}</Wrapper>
}

export default List
