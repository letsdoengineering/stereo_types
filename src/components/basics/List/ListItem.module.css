.is-links {
    color: var(--default-blue);
}
