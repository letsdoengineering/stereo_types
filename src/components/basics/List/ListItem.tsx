import React from 'react'
import getClassNames from '../../../utils/get-class-names'
import listItemStyles from './ListItem.module.css'

type ListItemProps = React.AllHTMLAttributes<HTMLLIElement> & {
  /** unique key for each <li> rendered to help react track dynamically rendered contents */
  listKey: string
  /** optional: styles a list to appear as links if set - defaults to false */
  isLinks?: boolean
}

/** Text: renders a Text tag of a type based on the tagName prop */
const ListItem: React.FC<ListItemProps> = ({
  listKey,
  className,
  isLinks = false,
  children,
  ...rest
}: ListItemProps) => {
  const classNames = getClassNames({
    className,
    conditionalClasses: { [`${listItemStyles.isLinks}`]: isLinks },
  })

  return (
    <li key={listKey} className={classNames} {...rest}>
      {children}
    </li>
  )
}

export default ListItem
