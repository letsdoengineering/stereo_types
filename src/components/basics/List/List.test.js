import * as React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import List from './List'
import renderer from 'react-test-renderer'
import { axe } from 'jest-axe'

describe('List', () => {
    describe('ordered prop', () => {
        it('renders ul type list by default', async () => {
            render(<List>a list</List>)
            const list = screen.getByText('a list')

            await waitFor(() => expect(list.nodeName).toBe('UL'))
        })

        it('renders ol type list if isOrdered is set to true', async () => {
            render(<List isOrdered>a list</List>)
            const text = screen.getByText('a list')

            await waitFor(() => expect(text.nodeName).toBe('OL'))
        })
    })

    it('has not changed the JSX layout of the page from previous snapshot', () => {
        const tree = renderer.create(<List />).toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders a list that is accessible', async () => {
        const { container } = render(<List />)
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})
