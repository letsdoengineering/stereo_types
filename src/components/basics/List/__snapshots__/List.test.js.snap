// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List has not changed the JSX layout of the page from previous snapshot 1`] = `
<ul
  className="listUnstyled"
/>
`;
